File: duna-core/src/main/java/io/duna/core/eventbus/event/DefaultEvent.java
package io.duna.core.eventbus.event;

import io.duna.core.concurrent.future.Future;
import io.duna.core.eventbus.DefaultEventBus;
import io.duna.core.eventbus.Event;
import io.duna.core.eventbus.message.Message;
import io.duna.core.function.Handler;
import io.reactivex.Observable;
import org.eclipse.collections.api.multimap.MutableMultimap;
import org.eclipse.collections.impl.factory.Lists;
import org.eclipse.collections.impl.factory.Multimaps;

import java.util.Arrays;
import java.util.List;
import java.util.UUID;
import java.util.function.Predicate;

public class DefaultEvent<T> implements Event<T> {

    private DefaultEventBus eventBus;

    private String name;

    private MutableMultimap<String, String> headers;

    private List<Predicate<? super T>> filters;

    private Handler<T> deadLetterHandler;

    public DefaultEvent(DefaultEventBus eventBus, String name) {
        this.headers = Multimaps.mutable.set.empty();
        this.filters = Lists.mutable.empty();
        this.eventBus = eventBus;
        this.name = name;
    }

    @Override
    public Event<T> header(String key, String value) {
        headers.put(key, value);
        return this;
    }

    @Override
    public Event<T> header(String key, String... values) {
        Arrays.stream(values)
            .forEach(v -> headers.put(key, v));
        return this;
    }

    @Override
    public Event<T> filter(Predicate<? super T> predicate) {
        filters.add(predicate);
        return this;
    }

    @Override
    public Event<T> deadLetterHandler(Handler<T> handler) {
        this.deadLetterHandler = handler;
        return this;
    }

    @Override
    public String getName() {
        return name;
    }

    @Override
    public Future<Void> purge() {
        return null;
    }

    @Override
    public Future<Void> purge(Handler<?> handler) {
        return null;
    }

    @Override
    public <V> Future<Message<V>> send(T message) {
        Future<Message<V>> future = null; // TODO put implementation here
        String address = UUID.randomUUID().toString();

        return future;
    }

    @Override
    public Future<Void> publish(T message) {
        return null;
    }

    @Override
    public Future<Void> enqueue(T message) {
        return null;
    }

    @Override
    public void consume(Handler<Message<T>> handler) {
//        eventBus.getEventRouter().<Message<T>>registerHandler(name, message -> {
//            if (filters.isEmpty()) {
//                handler.handle(message);
//            } else {
//                AtomicBoolean shouldExecute = new AtomicBoolean(true);
//                filters.forEach(p -> {
//                    //noinspection unchecked
//                    if (!p.test((T) message))
//                        shouldExecute.set(false);
//                });
//
//                if (shouldExecute.get()) {
//                    handler.handle(message);
//                }
//            }
//        });
    }

    @Override
    public void dequeue(Handler<Message<T>> handler, long interval) {

    }

    @Override
    public Observable<T> observe() {
        return null;
    }
}

File: duna-core/src/main/kotlin/io/duna/core/eventbus/event/DefaultEvent.kt
package io.duna.core.eventbus.event

import io.duna.core.Context
import io.duna.core.eventbus.EventBus
import io.duna.core.eventbus.Headers
import io.duna.core.eventbus.Message
import io.duna.core.eventbus.impl.MessageImpl
import org.eclipse.collections.api.multimap.Multimap
import org.eclipse.collections.impl.factory.Multimaps
import java.util.*
import java.util.function.Consumer

internal open class DefaultEvent<T>(private val eventBus: EventBus,
                                    override val name: String) : Event<T> {

  private val headers: Headers = Multimaps.mutable.set.empty()

  protected var deadLetterSink: Consumer<in Message<T>>? = null
    private set

  override fun publish(attachment: T?): Event<T> {
    dispatch(attachment)
    return this
  }

  override fun <V> request(attachment: T?): Subscriber<V> {
    return ResponseAwareSubscriber(eventBus, "uuid:${UUID.randomUUID()}") {
      dispatch(attachment, this@ResponseAwareSubscriber.event)
    }
  }

  override fun <V> poll(): Subscriber<V> {
    return ResponseAwareSubscriber(eventBus, "uuid:${UUID.randomUUID()}") {
      dispatch(null, this@ResponseAwareSubscriber.event)
    }
  }

  override fun onDeadLetter(deadLetterSink: Consumer<in Message<T>>): Event<T> {
    this.deadLetterSink = deadLetterSink
    return this
  }

  override fun putHeader(key: String, value: String): Event<T> {
    headers[key] += value
    return this
  }

  override fun putHeader(key: String, vararg values: String): Event<T> {
    headers[key] += values
    return this
  }

  override fun putHeader(values: Multimap<String, String>): Event<T> {
    headers.putAll(values)
    return this
  }

  @Suppress("UNCHECKED_CAST")
  private fun dispatch(attachment: Any?,
                       responseEvent: String? = null) {
    val message = if (attachment is Throwable)
      MessageImpl(
        source = (Context.Companion.currentContext()?.get("duna:current-subscriber") as Subscriber<*>?)?.event,
        target = name,
        responseEvent = responseEvent,
        attachment = attachment,
        headers = headers
      )
    else
      MessageImpl(
        source = (Context.Companion.currentContext()?.get("duna:current-subscriber") as Subscriber<*>?)?.event,
        target = name,
        responseEvent = responseEvent,
        attachment = attachment as T,
        headers = headers
      )

    eventBus
      .dispatch(message)
      .onError { deadLetterSink?.accept(message as Message<T>) }
  }
}
